freedink 108.4 / freedink-data 1.08.20170409

djpadz opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

bfontaine commented on Formula/freedink-data.rb:
```diff
@@ -0,0 +1,18 @@
+    system "sed", "-i.bak", "-e", "s,^VERSION,#VERSION,", "-e", "s,^SOURCE_DATE_EPOCH,#SOURCE_DATE_EPOCH,", "-e", "s,xargs -0r,xargs -0,", "Makefile"
+    version=Dir.pwd.split("/")[-1].split("-")[-1]
+    source_date = Dir.pwd.split("/")[-1].split(".")[-1].match(/(\d{4})(\d{2})(\d{2})/).to_a[1..-1].join("-")
+    source_date_epoch = Time.parse("#{source_date} 0:00 UTC").to_i
```
Why do we need to do all this parsing?

djpadz replied:
Believe me, I wasn't happy about this, either.  The author puts the build date into the version, which would lead to different people getting different versions of the code, depending on when they installed it.  I decided to rewrite it, so that the version would reflect the author's release date, rather than the build date.

bfontaine replied:
You should report this upstream.

bfontaine commented on Formula/freedink-data.rb:
```diff
@@ -0,0 +1,18 @@
+class FreedinkData < Formula
+  desc "Data for FreeDink."
```
You can make this a resource of the `freedink` formula instead of a formula in itself.

djpadz replied:
I'm not sure how to do that.  Can you send me a link to some documentation?

bfontaine replied:
See e.g. [`cracklib.rb`](https://github.com/Homebrew/homebrew-core/blob/312a38050f7bf3fa4f644a9d6d23d2d9dabac506/Formula/cracklib.rb#L18). You have to declare the external resource in a `resource do` block:

```ruby
resource "freedink-data" do
  url "..."
  sha256 "..."
end
```

Then in your `def install` you can unpack it with `Resource#stage`:
```ruby
resource("freedink-data").stage do
  # execute some code here to install the additional resource
  # the freedink-data resource is unpacked in a temporary
  # directory and your code is run in that directory.
end
```

bfontaine commented on Formula/freedink.rb:
```diff
@@ -0,0 +1,32 @@
+  end
+
+  test do
+    FileTest.executable? "src/freedink"
```
Please write a test that executes something instead of just checking if a file exists.

bfontaine commented on Formula/freedink.rb:
```diff
@@ -0,0 +1,32 @@
+                          "--prefix=#{prefix}"
+
+    system "make", "install"
+    ln_s "#{HOMEBREW_PREFIX}/freedink-data/share/dink", "#{share}/dink"
```
This should be done as a post-install step (in `def post_install`) but making `freeding-data` a resource of this formula is simpler.

djpadz replied:
I'm not quite sure I understand the second part of that.  How would I do that?

bfontaine replied:
See my other comment: https://github.com/Homebrew/homebrew-core/pull/12129#discussion_r114116644

bfontaine commented on Formula/freedink.rb:
```diff
@@ -0,0 +1,36 @@
+
+  test do
+    FileTest.executable? "src/freedink" &&
+    %x[src/freedink -vwis].split("\n").first == 'GNU FreeDink 108.4'
```
Please use `shell_output("…")` instead of `%x[...]`. Also, `freedink` should be installed in `bin/` if it’s executable.

djpadz commented:
Sorry I let this go so long...  Competing priorities...

ilovezfs commented:
@BrewTestBot test this please

djpadz commented:
@bfontaine I think I've addressed all of the issues.  The upstream issue isn't really one that's blocking the formula; more of a style decision on the part of the author, which I decided to work around.

bfontaine commented:
Please report the issue upstream anyway; it’d be better if they could provide a way to do so instead of us relying of this code.

(sorry for the late response, I was in holidays away from my computer)

djpadz commented:
I dropped the following email to `freedink-bugs@gnu.org`:

> Hi there-
> 
> I’m building a package for freedink and freedink-data for HomeBrew, and I received the following feedback from the maintainers, which I thought I’d pass along.
> 
> In the Makefile for freedink-data, the build date is included in the version string.  This is probably not the best idea, since it means that my version of freedink-data will appear to be newer than yours, simply by virtue of the fact that I happened to type make after you did.  It might be a good idea to set your versions to static strings.  If you want the build date in there somewhere, then put it elsewhere, but part of the version string probably isn’t the best place.
> 
> In any case, for HomeBrew, I’m munging the date portion of the source directory, and using that in lieu of a build date.
> 
> Best,
> 
> —Dj
> 

I've also updated the `freedink-data` resource to the latest version.

ilovezfs commented:
The current test block gets stuck on CI indefinitely until the build is aborted.

djpadz commented:
Found the cause of the test hang.  Should be okay now.

djpadz commented:
@bfontaine I got an email back from the developer, and, as it turns out, I didn't look at the `Makefile` hard enough.  The parsing stuff wasn't necessary, so I've removed that stuff.  All that's needed is the change from `xargs -0r` to `xargs -0`.

dunn commented on Formula/freedink.rb:
```diff
@@ -0,0 +1,39 @@
+
+    system "make", "install"
+    resource("freedink-data").stage do
+      system "sed", "-i.bak", "-e", "s,xargs -0r,xargs -0,", "Makefile"
```
We have an `inreplace` method that can be used instead of `sed`.  But this should also be reported upstream.

djpadz replied:
I mentioned the `sed` statement to the author. I haven't heard back, so I'd say we can file that under "lesser sins."  As for the `inreplace` method, I'll get a look at it, and push another commit later tonight.

dunn commented on Formula/freedink.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+
+  test do
+    FileTest.executable?("#{bin}/freedink") && shell_output("#{bin}/freedink -vwis").split("\n").first == "GNU FreeDink 108.4" && FileTest.exists?("#{share}/dink/dink/Dink.dat")
```
These can all be on separate lines; if any fail the test fails.

djpadz replied:
I don't think that's the case.  In a ruby code block, the return value is the last value in the block.  For example, I changed the `test do` block to this:

```
  test do
    FileTest.exists?("/foo/bar/baz")
    FileTest.executable?("#{bin}/freedink")
    shell_output("#{bin}/freedink -vwis").split("\n").first == "GNU FreeDink 108.4"
    FileTest.exists?("#{share}/dink/dink/Dink.dat")
  end
```

This test passes (you can try it yourself), even though the first test fails, because the block returns `true`.

A viable alternative would be:

```
  test do
    return false unless FileTest.executable?("#{bin}/freedink")
    return false unless shell_output("#{bin}/freedink -vwis").split("\n").first == "GNU FreeDink 108.4"
    return false unless FileTest.exists?("#{share}/dink/dink/Dink.dat")
  end
```

djpadz commented:
_Moving my response outside of the obsolete block_

@dunn I don't think that's the case.  In a ruby code block, the return value is the last value in the block.  For example, I changed the `test do` block to this:

```
  test do
    FileTest.exists?("/foo/bar/baz")
    FileTest.executable?("#{bin}/freedink")
    shell_output("#{bin}/freedink -vwis").split("\n").first == "GNU FreeDink 108.4"
    FileTest.exists?("#{share}/dink/dink/Dink.dat")
  end
```

This test passes (you can try it yourself), even though the first test fails, because the block returns `true`.

A viable alternative would be:

```
  test do
    return false unless FileTest.executable?("#{bin}/freedink")
    return false unless shell_output("#{bin}/freedink -vwis").split("\n").first == "GNU FreeDink 108.4"
    return false unless FileTest.exists?("#{share}/dink/dink/Dink.dat")
  end
```

dunn commented:
My mistake, sorry.  But instead of returns, do `assert File.exist? file` etc.

djpadz commented:
That seems reasonable :-)

bfontaine commented on Formula/freedink.rb:
```diff
@@ -0,0 +1,41 @@
+
+  test do
+    assert FileTest.executable?("#{bin}/freedink")
+    assert shell_output("#{bin}/freedink -vwis").split("\n").first == "GNU FreeDink 108.4"
```
Use `assert_match "GNU FreeDink 108.4", shell_output("#{bin}/freedink -vwis")` and remove the `FileTest.executable?` above; `shell_output` will fail if it’s not executable anyway.

djpadz replied:
Done.

bfontaine commented on Formula/freedink.rb:
```diff
@@ -34,8 +34,7 @@ def install
   test do
-    assert FileTest.executable?("#{bin}/freedink")
-    assert shell_output("#{bin}/freedink -vwis").split("\n").first == "GNU FreeDink 108.4"
+    assert_match "GNU FreeDink 108.4", shell_output("#{bin}/freedink -vwis").split("\n").first
```
You can remove `.split("\n").first`; `assert_match` already ensures `GNU FreeDink 108.4` is *somewhere* in that output. We have `assert_equal` for strict equality.

djpadz commented:
@bfontaine Done.

djpadz commented:
@bfontaine @dunn So, are you waiting on me for anything?  I believe we've cleared the upstream issue.

bfontaine commented:
I stepped away from maintainership since you opened the issue so you’ll have to wait for @dunn or others :wink:

dunn commented:
Merged as 89b2d82defa415f8106b25010a389119760d198b, thanks again!

djpadz commented:
Thanks, @dunn & @bfontaine !
